### PR TITLE
feat: allow authenticated + ssl mail

### PR DESCRIPTION
**Pull request by itkovian**

### CODE REVIEW

File: lib/vsc/utils/mail.py [hunk 1]
```diff
@@ -68,8 +69,25 @@ def __init__(self, mail_host=None, mail_to=None, mail_from=None, mail_subject=No
 class VscMail(object):
     """Class providing functionality to send out mail."""
 
-    def __init__(self, mail_host=None):
-        self.mail_host = mail_host
+    def __init__(self, mail_host=None, auth=False, mail_config=None):
+
+        opts = {
+            "from": ("Sender", None, "store", None),
+            "smtp": ("Host:Port", None, "store", "exploud.ugent.be:587"),
```
Comment (stdweird) [inline lib/vsc/utils/mail.py:76]: an ugent server as default in vsc-base?

File: lib/vsc/utils/mail.py [hunk 2]
```diff
@@ -68,8 +69,25 @@ def __init__(self, mail_host=None, mail_to=None, mail_from=None, mail_subject=No
...
+            "smtp-auth-password": ("Authentication password", None, "store", None),
+            "smtp-auth-user": ("Authentication user", None, "store", None),
+            "smtp-use-starttls": ("Should we start a TLS session", bool, "store_true", False),
+        }
+
+        self.auth = auth
+
+        if self.auth:
+            self.go = simple_option(opts, config_files=mail_config)
```
Comment (stdweird) [inline lib/vsc/utils/mail.py:85]: i don't like this tbh. this should be called outside this class

File: lib/vsc/utils/mail.py [hunk 3]
```diff
@@ -68,8 +69,25 @@ def __init__(self, mail_host=None, mail_to=None, mail_from=None, mail_subject=No
...
+            (host, port) = self.go.options.smtp.split(":")
+            self.mail_host = host
```
Comment (stdweird) [inline lib/vsc/utils/mail.py:87]: it's better to set these always, and default for host can be `''` and port `0`. it also simplifies the code below

File: lib/vsc/utils/mail.py [hunk 4]
```diff
@@ -85,13 +103,20 @@ def _send(self,
         """
 
         try:
-            if self.mail_host:
-                logging.debug("Using %s as the mail host", self.mail_host)
-                s = smtplib.SMTP(self.mail_host)
+            if self.auth:
+                logging.debug("Using %s as the mail host with authenticatio", self.mail_host)
+                s = smtplib.SMTP(host=self.mail_host, port=self.mail_port)
```
Comment (stdweird) [inline lib/vsc/utils/mail.py:108]: so this call can be put outside the if/else block when using default `''` and `0`

File: lib/vsc/utils/mail.py [hunk 5]
```diff
@@ -85,13 +103,20 @@ def _send(self,
...
-                logging.debug("Using the default mail host")
-                s = smtplib.SMTP()
-                s.connect()
+                if self.mail_host:
+                    logging.debug("Using %s as the mail host", self.mail_host)
+                    s = smtplib.SMTP(self.mail_host)
+                else:
+                    logging.debug("Using the default mail host")
+                    s = smtplib.SMTP()
+                    s.connect()
```
Comment (stdweird) [inline lib/vsc/utils/mail.py:119]: is the indent correct here? no connect if mail_host is configured?

File: lib/vsc/utils/mail.py [hunk 6]
```diff
@@ -105,6 +130,10 @@ def _send(self,
                 raise
             except smtplib.SMTPDataError as err:
                 raise
+
+            if self.auth:
```
Comment (stdweird) [inline lib/vsc/utils/mail.py:134]: what does this do? (and in particular, why is it only needed / allowed if auth is used?)

**Comment by boegel** [comment]
Please make sure the changes you're making are covered by the tests.

File: lib/vsc/utils/mail.py (hunk 1 above)
Comment (itkovian) [reply lib/vsc/utils/mail.py:76]: Good point.

**Comment by itkovian** [comment]
I thought a bit about it and I am not too happy with it either.

File: lib/vsc/utils/mail.py (hunk 5 above)
Comment (itkovian) [reply lib/vsc/utils/mail.py]: It was like this forever, but I'm not sure now.

File: lib/vsc/utils/mail.py [hunk 7]
```diff
@@ -68,8 +67,45 @@ def __init__(self, mail_host=None, mail_to=None, mail_from=None, mail_subject=No
 class VscMail(object):
     """Class providing functionality to send out mail."""
 
-    def __init__(self, mail_host=None):
-        self.mail_host = mail_host
+    def __init__(self, mail_host=None, mail_port=None, smtp_auth_user=None, smtp_auth_password=None, smtp_use_starttls=False):
```
Comment (stdweird) [inline lib/vsc/utils/mail.py:70]: defaults for host and port can be  `''` and `0`

File: lib/vsc/utils/mail.py [hunk 8]
```diff
@@ -68,8 +67,45 @@ def __init__(self, mail_host=None, mail_to=None, mail_from=None, mail_subject=No
...
+        self.mail_host = host
+        self.mail_port = port
+        self.smtp_auth_user = smtp_auth_user
+        self.smtp_auth_password = smtp_auth_password
+        self.smtp_use_starttls = smtp_use_starttls
+
+    def _connect(self):
+
+        logging.debug("Using %s as the mail host with authenticatio", self.mail_host)
+        if self.mail_host:
```
Comment (stdweird) [inline lib/vsc/utils/mail.py:90]: use the defaults

File: lib/vsc/utils/mail.py [hunk 9]
```diff
@@ -68,8 +67,45 @@ def __init__(self, mail_host=None, mail_to=None, mail_from=None, mail_subject=No
...
 
-    def __init__(self, mail_host=None):
-        self.mail_host = mail_host
+    def __init__(self, mail_host=None, mail_port=None, smtp_auth_user=None, smtp_auth_password=None, smtp_use_starttls=False):
+
+        # we must not have the port in the host as this will fail the certificate check if needed
+        (host, port) = (mail_host, mail_port)
+        if mail_host:
+            try:
+                (host, port_) = mail_host.split(":")
```
Comment (stdweird) [inline lib/vsc/utils/mail.py:76]: why is the even allowed?

Comment (itkovian) [reply lib/vsc/utils/mail.py:76]: why is the what?

Comment (stdweird) [reply lib/vsc/utils/mail.py:76]: the format with a `:` in it

Comment (itkovian) [reply lib/vsc/utils/mail.py:82]: Dunno, somebody passed `explout.ugent.be:587` to it at some point :p

File: lib/vsc/utils/mail.py [hunk 10]
```diff
@@ -68,14 +67,57 @@ def __init__(self, mail_host=None, mail_to=None, mail_from=None, mail_subject=No
...
+
+        if self.smtp_auth_user and self.smtp_auth_password:
+            s.login(user=self.smtp_auth_user, password=self.smtp_auth_password)
+
+        s.connect()
+
+        return s
 
     def _send(self,
-              mail_from,
```
Comment (stdweird) [inline lib/vsc/utils/mail.py:75]: why was this needed? it looks horrible

File: lib/vsc/utils/mail.py [hunk 11]
```diff
@@ -68,14 +67,57 @@ def __init__(self, mail_host=None, mail_to=None, mail_from=None, mail_subject=No
...
+            s = smtplib.SMTP()
+
+        if self.smtp_use_starttls:
+            context = ssl.create_default_context()
+            s.starttls(context=context)
+
+        if self.smtp_auth_user and self.smtp_auth_password:
+            s.login(user=self.smtp_auth_user, password=self.smtp_auth_password)
+
+        s.connect()
```
Comment (stdweird) [inline lib/vsc/utils/mail.py:112]: should check the code again to introduce `s.quit` again

File: lib/vsc/utils/mail.py (hunk 9 above)
Comment (itkovian) [reply lib/vsc/utils/mail.py:82]: ah no, my bad, the config has `set smtp=exploud.ugent.be:587`

File: lib/vsc/utils/mail.py [hunk 12]
```diff
@@ -68,14 +67,49 @@ def __init__(self, mail_host=None, mail_to=None, mail_from=None, mail_subject=No
...
-    def __init__(self, mail_host=None):
+    def __init__(
+        self,
+        mail_host='',
+        mail_port=0,
+        smtp_auth_user=None,
+        smtp_auth_password=None,
+        smtp_use_starttls=False):
+
+        # we must not have the port in the host as this will fail the certificate check if needed
```
Comment (stdweird) [inline lib/vsc/utils/mail.py:78]: comment makes no sense anymore?

File: lib/vsc/utils/mail.py [hunk 13]
```diff
@@ -68,14 +67,49 @@ def __init__(self, mail_host=None, mail_to=None, mail_from=None, mail_subject=No
...
         self.mail_host = mail_host
+        self.mail_port = mail_port
+        self.smtp_auth_user = smtp_auth_user
+        self.smtp_auth_password = smtp_auth_password
+        self.smtp_use_starttls = smtp_use_starttls
+
+    def _connect(self):
+
```
Comment (stdweird) [inline lib/vsc/utils/mail.py:86]: docstring

File: lib/vsc/utils/mail.py (hunk 8 above)
Comment (stdweird) [reply lib/vsc/utils/mail.py:90]: please use the defaults

File: lib/vsc/utils/mail.py (hunk 11 above)
Comment (stdweird) [reply lib/vsc/utils/mail.py:111]: how was this resolved?

File: lib/vsc/utils/mail.py [hunk 14]
```diff
@@ -68,14 +67,57 @@ def __init__(self, mail_host=None, mail_to=None, mail_from=None, mail_subject=No
...
+        if self.smtp_auth_user and self.smtp_auth_password:
+            s.login(user=self.smtp_auth_user, password=self.smtp_auth_password)
+            logging.debug("Authenticated")
+
+        s.connect()
+
+        return s
+
+    def _send(
```
Comment (wdpypere) [inline lib/vsc/utils/mail.py:115]: can you put the function defenition on one line?

File: lib/vsc/utils/mail.py [hunk 15]
```diff
@@ -105,27 +142,33 @@ def _send(self,
...
+            raise VscMailError(
+                mail_host=self.mail_host,
+                mail_to=mail_to,
+                mail_from=mail_from,
+                mail_subject=mail_subject,
+                err=err)
+        finally:
+            s.quit()
+
+    def sendTextMail(
```
Comment (wdpypere) [inline lib/vsc/utils/mail.py:165]: same here?

File: lib/vsc/utils/mail.py (hunk 11 above)
Comment (stdweird) [reply lib/vsc/utils/mail.py:111]: ok i missed the finally block

File: lib/vsc/utils/mail.py (hunk 8 above)
Comment (itkovian) [reply lib/vsc/utils/mail.py:90]: I went through the code again, defaults are used when there is nothing provided, so I am not sure I can follow here.

Comment (stdweird) [reply lib/vsc/utils/mail.py:90]: i meant to replace the whole if/else with
```python
logging.debug("Using %s as the mail host and %d as mail port", self.mail_host, self.mail_port)
s = smtplib.SMTP(host=self.mail_host, port=self.mail_port)
```

**Comment by boegel** [comment]
Merged without version bump?
